File: python_solutions/boj_4949.py
allow=['(',')','[',']']
while(True):
  stack=[True]
  line=input()
  string=[char for char in line if (char in allow)]
  for char in string:
    if(char=='['):stack.append('[')
    elif(char==']' and stack[-1]=='['):stack.pop(-1)
    elif(char=='('):stack.append('(')
    elif(char==')' and stack[-1]=='('):stack.pop(-1)
    else:
      stack.append(False)
      break
  if line=='.':break
  print('yes' if stack==[True] else 'no')

File: cpp_solutions/boj_4949/boj_4949.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    while(1){
        string ss; getline(cin,ss);
        if (ss == ".") break;

        stack<char> st; bool ret = true;

        for(char c : ss){
            if (c == '(') st.push('(');
            else if (c == '[') st.push('[');
            else if (c == ')') {
                if (st.empty() || st.top() != '('){
                    ret = false;
                    break;
                }
                st.pop();
            } else if (c == ']'){
                if (st.empty() || st.top() != '['){
                    ret = false;
                    break;
                }
                st.pop(); 
            }
        }

        if (st.empty() && ret) cout << "yes\n";
        else cout << "no\n";
    }
}